walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+   * Returns whether a merge has ocurred or not and modifies the Collection of TimeRanges to include
+   * a merged TimeRange where the newTimeRange overlaps with the passed Collection.
+   */
+  private boolean mergeOnOverlap(Collection<TimeRange> timeRanges, TimeRange newTimeRange) {
+    // The default value is set to false, which is converted to true if a merge takes place.
+    boolean doOverlap = false;
+
+    // Store items that we will add and remove from the collection after we finish iterating
+    // through it in order to avoid ConcurrentModificationException from being thrown.
+    List<TimeRange> toRemove = new ArrayList<>();
+    List<TimeRange> toAdd = new ArrayList<>();
+
+    for (TimeRange takenTimeRange : timeRanges) {
+      if (newTimeRange.overlaps(takenTimeRange)) {
+        TimeRange combinedTimeRange = mergeTimeRanges(newTimeRange, takenTimeRange);
+        toRemove.add(takenTimeRange);
+        toAdd.add(combinedTimeRange);
+        doOverlap = true;
+      }
+    }
+
+    for (TimeRange timeRange : toRemove) {
+      timeRanges.remove(timeRange);
+    }
+
+    for (TimeRange timeRange : toAdd) {
+      timeRanges.add(timeRange);
+    }
+
+    return doOverlap;
+  }
+
+  private TimeRange mergeTimeRanges(TimeRange firstTimeRange, TimeRange secondTimeRange) {
+    TimeRange mergedTimeRange =
+        TimeRange.fromStartEnd(
+            Math.min(firstTimeRange.start(), secondTimeRange.start()),
+            Math.max(firstTimeRange.end(), secondTimeRange.end()),
+            /* inclusive = */ false);
+
+    return mergedTimeRange;
+  }
+
+  private List<TimeRange> getAvailableTimeRanges(
+      List<TimeRange> unavailableTimeRanges, long durationMinutes) {
+    if (unavailableTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> availableTimeRanges = new ArrayList<>();
+
+    // Check for available time starting at the beginning of the day.
+    if (unavailableTimeRanges.get(0).start() >= durationMinutes) {
+      availableTimeRanges.add(
+          TimeRange.fromStartEnd(
+              TimeRange.WHOLE_DAY.start(),
+              unavailableTimeRanges.get(0).start(),
+              /* inclusive = */ false));
+    }
+
+    for (int i = 0; i < unavailableTimeRanges.size() - 1; i++) {
+      if (enoughTimeBetween(
+          unavailableTimeRanges.get(i), unavailableTimeRanges.get(i + 1), durationMinutes)) {
+        availableTimeRanges.add(
+            TimeRange.fromStartEnd(
+                unavailableTimeRanges.get(i).end(),
+                unavailableTimeRanges.get(i + 1).start(),
+                /* inclusive = */ false));
+      }
+    }
+
+    // Check for available time at the end of the day.
+    int timeAfterLastMeeting =
+        TimeRange.WHOLE_DAY.end()
+            - unavailableTimeRanges.get(unavailableTimeRanges.size() - 1).end();
```

Comment:
Maybe extract `unavailableTimeRanges.get(unavailableTimeRanges.size() - 1)` our into an intermediate variable since you use it a couple times.

Reply:
Done.